plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunction.java

```diff
@@ -0,0 +1,170 @@
+        DeltaLakeMetadata deltaLakeMetadata = deltaLakeMetadataFactory.create(session.getIdentity());
+        DeltaLakeTableHandle tableHandle = deltaLakeMetadata.getTableHandle(session, schemaTableName);
+        validateThatCdfWasEnabled(session, fileSystemFactory, tableHandle, sinceVersion);
+        List<DeltaLakeColumnHandle> columnHandles = deltaLakeMetadata.getColumnHandles(session, tableHandle)
+                .values().stream()
+                .map(columnHandle -> (DeltaLakeColumnHandle) columnHandle)
+                .filter(column -> !IGNORED_COLUMNS.contains(column.getName()))
+                .collect(toImmutableList());
+
+        ImmutableList.Builder<Descriptor.Field> outputFields = ImmutableList.builder();
+        columnHandles.stream()
+                .map(columnHandle -> new Descriptor.Field(columnHandle.getName(), Optional.of(columnHandle.getType())))
+                .forEach(outputFields::add);
+
+        outputFields.add(new Descriptor.Field(CHANGE_TYPE_COLUMN_NAME, Optional.of(VARCHAR)));
+        outputFields.add(new Descriptor.Field(COMMIT_VERSION_COLUMN_NAME, Optional.of(INTEGER)));
+        outputFields.add(new Descriptor.Field(COMMIT_TIMESTAMP_COLUMN_NAME, Optional.of(TIMESTAMP_TZ_MILLIS)));
+
+        return TableFunctionAnalysis.builder()
+                .handle(new TableChangesFunctionTableHandle(schemaName, tableName, sinceVersion, columnHandles))
+                .returnedType(new Descriptor(outputFields.build()))
+                .build();
+    }
+
+    public void validateThatCdfWasEnabled(
+            ConnectorSession session,
+            TrinoFileSystemFactory fileSystemFactory,
+            DeltaLakeTableHandle tableHandle,
+            long sinceVersion)
+    {
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        Path transactionLogDir = getTransactionLogDir(new Path(tableHandle.getLocation()));
+        long currentVersion = sinceVersion;
+
+        try {
+            while (currentVersion < tableHandle.getReadVersion()) {
+                Optional<List<DeltaLakeTransactionLogEntry>> entries = getEntriesFromJson(currentVersion, transactionLogDir, fileSystem);
+                if (entries.isPresent()) {
+                    boolean containsRemoveEntry = false;
+                    boolean containsCdcEntry = false;
+                    for (DeltaLakeTransactionLogEntry entry : entries.get()) {
+                        if (entry.getCDC() != null) {
+                            containsCdcEntry = true;
+                        }
+                        if (entry.getRemove() != null) {
+                            containsRemoveEntry = true;
+                        }
+                    }
+                    if (containsRemoveEntry && !containsCdcEntry) {
+                        throw new TrinoException(DELTA_LAKE_BAD_DATA, "Change Data Feed is not enabled for the requested versions range");
```

Comment:
Why `DELTA_LAKE_BAD_DATA` ? This is rather a user problem of requesting content on a table not having CDF enabled, and not the table being incorrectly filled.

Reply:
Did you omit changing the error code ?